### ISSUE TITLE
mypy doesn't understand function signature compatibility

### ISSUE THREAD

**Issue by GBeauregard** [python/mypy#11807]
**Bug Report**

mypy doesn't understand the signature compatibility between functions when determining valid uses in lists

**To Reproduce**
The following code type checks despite containing a type error that will lead to runtime failure. The expectation was that appending to the list will show a type error or warning for the incompatible signatures.
```python
funclist = [lambda x: print(1)]
funclist += [lambda k: print(2)]  # incompatible signature, but type not widened and no error
reveal_type(funclist)  # builtins.list[def (x: Any)]
funclist[1](x=4)  # this will error at runtime because of a typing mistake
```

**Expected Behavior**
Type error. I compared to pyright and pyright does show a type error for the incompatible signatures. Other possible behaviors would be to widen the type of `funclist` on line 2 to no longer think the first argument has keyword variable named `x`, but this is probably a bit odd/inconsistent. Normally you would need to annotate `funclist` when it's defined if you wanted the wider behavior.

**Actual Behavior**
no type error

(Write what happened.)

**Your Environment**

- Mypy version used: 0.920
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.10
- Operating system and version: Gentoo/Linux stable

**Comment by DanielNoord** [comment]
Duplicate of https://github.com/python/mypy/issues/1013.
